lib/active_shipping/carriers/australia_post.rb

```diff
@@ -0,0 +1,249 @@
+    end
+
+    def calculate_rates(origin, destination, packages, service_code, options = {})
+      packages = Array(packages)
+
+      calculate_requests = packages.map do |package|
+        calculate_request = CalculateRequest.new(origin, destination, package, service_code, options)
+
+        calculate_request.parse(commit(calculate_request.url))
+        calculate_request
+      end
+
+      combined_response = CombinedResponse.new(origin, destination, packages, calculate_requests)
+
+      RateResponse.new(true, 'success', combined_response.params, combined_response.options)
+    end
+
+    private
+
+    def commit(request_url)
+      ssl_get(request_url, headers)
+
+    rescue ActiveUtils::ResponseError, ActiveShipping::ResponseError => e
+      data = JSON.parse(e.response.body)
+
+      RateResponse.new(false, data['error']['errorMessage'], data)
+    end
+
+    def headers
+      {
+        'Content-type' => 'application/json',
+        'auth-key'     => @options[:api_key]
+      }
+    end
+
+    class CombinedResponse
+
+      def initialize(origin, destination, packages, requests)
+        @requests    = requests
+        @origin      = origin
+        @destination = destination
+        @packages    = packages
+      end
+
+      def options
+        {
+          rates:         rates,
+          raw_responses: @requests.map(&:raw_response),
+          request:       @requests.map(&:url)
+        }
+      end
+
+      def params
+        {
+          responses: @requests.map(&:response)
+        }
+      end
+
+      private
+
+      def rate_options(rates)
+        {
+          service_name:       rates.first[:service_name],
+          service_code:       rates.first[:service_code],
+          total_price:        rates.sum { |rate| rate[:total_price] },
+          currency:           'AUD',
+          delivery_time_text: rates.first[:delivery_time_text]
+        }
+      end
+
+      def rates
+        rates = @requests.map(&:rates).flatten
+
+        rates.group_by { |rate| rate[:service_name] }.map do |service_name, service_rates|
+          next unless service_rates.size == @packages.size
+
+          AustraliaPostRateEstimate.new(@origin, @destination, AustraliaPost.name, service_name, rate_options(service_rates))
+        end.compact
+      end
+
+    end
+
+    class AustraliaPostRequest
+      attr_reader :raw_response
+      attr_reader :response
+      attr_reader :rates
+
+      def initialize(origin, destination, package, options)
+        @origin      = Location.from(origin)
+        @destination = Location.from(destination)
+        @package     = package
+        @rates       = []
+        @options     = options
+      end
+
+      def url
+        endpoint = domestic_destination? ? @endpoints[:domestic] : @endpoints[:international]
+        params   = domestic_destination? ? domestic_params : international_params
+
+        "#{URL}#{endpoint}.json?#{params.to_query}"
```

Comment:
What do you think about using [URI](http://ruby-doc.org/stdlib-2.1.2/libdoc/uri/rdoc/URI/HTTP.html) here? 

It's the kind of problem that appears simple, but is easy to reimplement incorrectly in some subtle way.